File: backend/seed.py
"""Script de seed: importa los 1324 ejercicios del dataset y crea
cuentas demo (entrenador + 2 alumnos) con un plan semanal y, para
alumno1, varias semanas de historial de entrenamiento.

Uso:
    cd backend
    source .venv/bin/activate
    python seed.py
"""

from __future__ import annotations

import json
from datetime import datetime, timedelta, timezone
from pathlib import Path

from sqlalchemy.orm import Session

from app.database import Base, SessionLocal, engine
from app.models import Exercise, PlanEntry, RoleEnum, SetLog, User, WeightUnitEnum, WorkoutSession
from app.security import hash_password
from app.utils import today_date_key

DATA_PATH = Path(__file__).parent / "data" / "exercises.seed.json"


def seed_exercises(db: Session) -> None:
    raw = json.loads(DATA_PATH.read_text(encoding="utf-8"))
    print(f"Importando {len(raw)} ejercicios...")

    existing_ids = {row[0] for row in db.query(Exercise.id).all()}
    created = 0
    for item in raw:
        if item["id"] in existing_ids:
            continue
        db.add(
            Exercise(
                id=item["id"],
                name=item["name"],
                search_name=item["name"].lower(),
                category=item["category"],
                body_part=item["body_part"],
                equipment=item["equipment"],
                muscle_group=item["muscle_group"],
                target=item["target"],
                secondary_muscles=json.dumps(item["secondary_muscles"], ensure_ascii=False),
                instructions_es=item["instructions_es"],
                instructions_en=item["instructions_en"],
                instruction_steps_es=json.dumps(item["instruction_steps_es"], ensure_ascii=False),
                image=item["image"],
                gif_url=item["gif_url"],
                attribution=item["attribution"],
            )
        )
        created += 1
    db.commit()
    print(f"Ejercicios importados ({created} nuevos, {len(existing_ids)} ya existían).")


def get_or_create_user(
    db: Session,
    *,
    email: str,
    name: str,
    password: str,
    role: RoleEnum,
    weight_unit: WeightUnitEnum,
) -> User:
    user = db.query(User).filter(User.email == email).first()
    if user:
        return user
    user = User(
        email=email,
        name=name,
        password_hash=hash_password(password),
        role=role,
        weight_unit=weight_unit,
    )
    db.add(user)
    db.commit()
    db.refresh(user)
    return user


def pick_exercises(db: Session, body_part: str, equipment: list[str], take: int) -> list[Exercise]:
    return (
        db.query(Exercise)
        .filter(Exercise.body_part == body_part, Exercise.equipment.in_(equipment))
        .order_by(Exercise.id.asc())
        .limit(take)
        .all()
    )


def seed_users_and_plan(db: Session) -> None:
    trainer = get_or_create_user(
        db,
        email="trainer@gymcristiano.app",
        name="Coach Cristiano",
        password="Entrenador123!",
        role=RoleEnum.TRAINER,
        weight_unit=WeightUnitEnum.KG,
    )
    print(f"Entrenador demo: {trainer.email} / Entrenador123!")

    alumno1 = get_or_create_user(
        db,
        email="alumno1@gymcristiano.app",
        name="Marco Torres",
        password="Alumno123!",
        role=RoleEnum.USER,
        weight_unit=WeightUnitEnum.KG,
    )
    alumno2 = get_or_create_user(
        db,
        email="alumno2@gymcristiano.app",
        name="Laura Gómez",
        password="Alumno123!",
        role=RoleEnum.USER,
        weight_unit=WeightUnitEnum.LB,
    )
    print(f"Alumnos demo: {alumno1.email} / Alumno123!  y  {alumno2.email} / Alumno123!")

    # Split clásico de 3 días: Lunes=Empuje, Miércoles=Tirón, Viernes=Piernas.
    push = pick_exercises(db, "chest", ["barbell", "dumbbell"], 2)
    push_shoulders = pick_exercises(db, "shoulders", ["barbell", "dumbbell"], 1)
    pull = pick_exercises(db, "back", ["barbell", "dumbbell", "cable"], 3)
    legs = pick_exercises(db, "upper legs", ["barbell", "dumbbell"], 3)

    plan: list[tuple[int, list[Exercise]]] = [
        (0, [*push, *push_shoulders]),  # Lunes
        (2, pull),  # Miércoles
        (4, legs),  # Viernes
    ]

    for target_user in (alumno1, alumno2):
        for day_of_week, exercises in plan:
            db.query(PlanEntry).filter(
                PlanEntry.user_id == target_user.id, PlanEntry.day_of_week == day_of_week
            ).delete()
            for i, exercise in enumerate(exercises):
                db.add(
                    PlanEntry(
                        user_id=target_user.id,
                        day_of_week=day_of_week,
                        exercise_id=exercise.id,
                        position=i,
                        sets=4,
                        reps=10,
                    )
                )
    db.commit()
    print("Plan semanal asignado a ambos alumnos.")

    # Historial de entrenamiento realista solo para alumno1, con peso
    # creciente semana a semana, para que el gráfico de progreso
    # muestre una tendencia visible al primer vistazo. alumno2 queda
    # con plan pero sin historial, para probar el estado "sin
    # entrenos todavía" en el panel del entrenador.
    today = datetime.now(timezone.utc).date()
    base_weights: dict[str, float] = {}

    for week in (3, 2, 1):
        for day_of_week, exercises in plan:
            days_back = week * 7 - day_of_week
            session_date = today - timedelta(days=days_back)
            date_key = today_date_key(session_date)

            session = (
                db.query(WorkoutSession)
                .filter(WorkoutSession.user_id == alumno1.id, WorkoutSession.date_key == date_key)
                .first()
            )
            if not session:
                session = WorkoutSession(
                    user_id=alumno1.id,
                    day_of_week=day_of_week,
                    date_key=date_key,
                    date=datetime.combine(session_date, datetime.min.time(), tzinfo=timezone.utc),
                    completed=True,
                )
                db.add(session)
                db.commit()
                db.refresh(session)

            for exercise in exercises:
                if exercise.id not in base_weights:
                    base_weights[exercise.id] = 40.0 if exercise.equipment == "barbell" else 12.0
                week_index = 3 - week  # 0, 1, 2
                weight_kg = base_weights[exercise.id] + week_index * 2.5

                for set_number in range(1, 5):
                    existing_set = (
                        db.query(SetLog)
                        .filter(
                            SetLog.session_id == session.id,
                            SetLog.exercise_id == exercise.id,
                            SetLog.set_number == set_number,
                        )
                        .first()
                    )
                    if not existing_set:
                        db.add(
                            SetLog(
                                session_id=session.id,
                                exercise_id=exercise.id,
                                set_number=set_number,
                                weight_kg=weight_kg,
                                reps=10,
                            )
                        )
    db.commit()
    print("Historial de entrenamiento generado para alumno1.")


def main() -> None:
    Base.metadata.create_all(bind=engine)
    db = SessionLocal()
    try:
        seed_exercises(db)
        seed_users_and_plan(db)
    finally:
        db.close()
    print("\n🌱 Seed completo.")


if __name__ == "__main__":
    main()

File: prisma/seed.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { PrismaClient } from "../src/generated/prisma/client";
import { PrismaNeonHTTP } from "@prisma/adapter-neon";
import { hashPassword } from "../src/lib/auth";
import { todayDateKey } from "../src/lib/constants";

// El seed usa el adapter HTTP (fetch puro, sin WebSocket) a propósito:
// corre desde entornos locales/CI que a veces no permiten salida
// WebSocket (solo HTTPS), y no necesita transacciones interactivas.
// La app en runtime (src/lib/prisma.ts) usa el pool WS (PrismaNeon)
// porque Cloudflare Workers sí lo soporta nativamente.
const adapter = new PrismaNeonHTTP(process.env.DATABASE_URL ?? "", {});
const prisma = new PrismaClient({ adapter });

type SeedExercise = {
  id: string;
  name: string;
  category: string;
  body_part: string;
  equipment: string;
  muscle_group: string;
  secondary_muscles: string[];
  target: string;
  instructions_es: string;
  instructions_en: string;
  instruction_steps_es: string[];
  image: string;
  gif_url: string;
  attribution: string;
};

async function seedExercises() {
  const dataPath = fileURLToPath(
    new URL("./data/exercises.seed.json", import.meta.url),
  );
  const raw = readFileSync(dataPath, "utf-8");
  const exercises: SeedExercise[] = JSON.parse(raw);

  console.log(`Importando ${exercises.length} ejercicios...`);
  for (const e of exercises) {
    await prisma.exercise.upsert({
      where: { id: e.id },
      create: {
        id: e.id,
        name: e.name,
        searchName: e.name.toLowerCase(),
        category: e.category,
        bodyPart: e.body_part,
        equipment: e.equipment,
        muscleGroup: e.muscle_group,
        secondaryMuscles: JSON.stringify(e.secondary_muscles),
        target: e.target,
        instructionsEs: e.instructions_es,
        instructionsEn: e.instructions_en,
        instructionStepsEs: JSON.stringify(e.instruction_steps_es),
        image: e.image,
        gifUrl: e.gif_url,
        attribution: e.attribution,
      },
      update: {
        name: e.name,
        searchName: e.name.toLowerCase(),
        category: e.category,
        bodyPart: e.body_part,
        equipment: e.equipment,
        muscleGroup: e.muscle_group,
        secondaryMuscles: JSON.stringify(e.secondary_muscles),
        target: e.target,
        instructionsEs: e.instructions_es,
        instructionsEn: e.instructions_en,
        instructionStepsEs: JSON.stringify(e.instruction_steps_es),
        image: e.image,
        gifUrl: e.gif_url,
        attribution: e.attribution,
      },
    });
  }
  console.log("Ejercicios importados.");
}

async function pickExercises(
  bodyPart: string,
  equipment: string[],
  take: number,
) {
  return prisma.exercise.findMany({
    where: { bodyPart, equipment: { in: equipment } },
    take,
    orderBy: { id: "asc" },
  });
}

function addDays(base: Date, days: number): Date {
  const d = new Date(base);
  d.setDate(d.getDate() + days);
  return d;
}

// El adapter HTTP (ver arriba) no soporta las transacciones internas
// que Prisma usa para `upsert()` en algunos modelos (falla con
// "Transactions are not supported in HTTP mode"). Todos los `update`
// de este seed son no-op de cualquier forma (solo nos interesa
// idempotencia al re-ejecutar), así que reemplazamos upsert por un
// find-o-create simple, que sí es un statement único.
async function findOrCreate<T>(
  find: () => Promise<T | null>,
  create: () => Promise<T>,
): Promise<T> {
  const existing = await find();
  return existing ?? create();
}

async function seedUsersAndPlan() {
  const trainerPassword = await hashPassword("Entrenador123!");
  const trainer = await findOrCreate(
    () => prisma.user.findUnique({ where: { email: "trainer@gymcristiano.app" } }),
    () =>
      prisma.user.create({
        data: {
          email: "trainer@gymcristiano.app",
          name: "Coach Cristiano",
          passwordHash: trainerPassword,
          role: "TRAINER",
        },
      }),
  );
  console.log(`Entrenador demo: ${trainer.email} / Entrenador123!`);

  const alumno1Password = await hashPassword("Alumno123!");
  const alumno1 = await findOrCreate(
    () => prisma.user.findUnique({ where: { email: "alumno1@gymcristiano.app" } }),
    () =>
      prisma.user.create({
        data: {
          email: "alumno1@gymcristiano.app",
          name: "Marco Torres",
          passwordHash: alumno1Password,
          role: "USER",
          weightUnit: "KG",
        },
      }),
  );

  const alumno2Password = await hashPassword("Alumno123!");
  const alumno2 = await findOrCreate(
    () => prisma.user.findUnique({ where: { email: "alumno2@gymcristiano.app" } }),
    () =>
      prisma.user.create({
        data: {
          email: "alumno2@gymcristiano.app",
          name: "Laura Gómez",
          passwordHash: alumno2Password,
          role: "USER",
          weightUnit: "LB",
        },
      }),
  );
  console.log(
    `Alumnos demo: ${alumno1.email} / Alumno123!  y  ${alumno2.email} / Alumno123!`,
  );

  // Split clásico de 3 días: Lunes=Empuje, Miércoles=Tirón, Viernes=Piernas.
  const push = await pickExercises("chest", ["barbell", "dumbbell"], 2);
  const pushShoulders = await pickExercises(
    "shoulders",
    ["barbell", "dumbbell"],
    1,
  );
  const pull = await pickExercises("back", ["barbell", "dumbbell", "cable"], 3);
  const legs = await pickExercises(
    "upper legs",
    ["barbell", "dumbbell"],
    3,
  );

  const plan: { dayOfWeek: number; exercises: typeof push }[] = [
    { dayOfWeek: 0, exercises: [...push, ...pushShoulders] }, // Lunes
    { dayOfWeek: 2, exercises: pull }, // Miércoles
    { dayOfWeek: 4, exercises: legs }, // Viernes
  ];

  for (const targetUser of [alumno1, alumno2]) {
    for (const day of plan) {
      await prisma.planEntry.deleteMany({
        where: { userId: targetUser.id, dayOfWeek: day.dayOfWeek },
      });
      let order = 0;
      for (const exercise of day.exercises) {
        await prisma.planEntry.create({
          data: {
            userId: targetUser.id,
            dayOfWeek: day.dayOfWeek,
            exerciseId: exercise.id,
            order: order++,
            sets: 4,
            reps: 10,
          },
        });
      }
    }
  }
  console.log("Plan semanal asignado a ambos alumnos.");

  // Historial de entrenamiento realista solo para alumno1, con peso
  // creciente semana a semana, para que ProgressChart muestre tendencia.
  const today = new Date();
  const weeksAgo = [3, 2, 1]; // 3 semanas de sesiones por cada día del split
  const baseWeights: Record<string, number> = {};

  for (const week of weeksAgo) {
    for (const day of plan) {
      // Aproxima la fecha real hacia atrás manteniendo separación de días.
      const daysBack = week * 7 - day.dayOfWeek;
      const date = addDays(today, -daysBack);
      const dateKey = todayDateKey(date);

      const session = await findOrCreate(
        () =>
          prisma.workoutSession.findUnique({
            where: { userId_dateKey: { userId: alumno1.id, dateKey } },
          }),
        () =>
          prisma.workoutSession.create({
            data: {
              userId: alumno1.id,
              dayOfWeek: day.dayOfWeek,
              dateKey,
              date,
              completed: true,
            },
          }),
      );

      for (const exercise of day.exercises) {
        if (!(exercise.id in baseWeights)) {
          baseWeights[exercise.id] =
            exercise.equipment === "barbell" ? 40 : 12;
        }
        const weekIndex = 3 - week; // 0, 1, 2
        const weightKg = baseWeights[exercise.id] + weekIndex * 2.5;

        for (let setNumber = 1; setNumber <= 4; setNumber++) {
          await findOrCreate(
            () =>
              prisma.setLog.findUnique({
                where: {
                  sessionId_exerciseId_setNumber: {
                    sessionId: session.id,
                    exerciseId: exercise.id,
                    setNumber,
                  },
                },
              }),
            () =>
              prisma.setLog.create({
                data: {
                  sessionId: session.id,
                  exerciseId: exercise.id,
                  setNumber,
                  weightKg,
                  reps: 10,
                },
              }),
          );
        }
      }
    }
  }
  console.log("Historial de entrenamiento generado para alumno1.");
}

async function main() {
  await seedExercises();
  await seedUsersAndPlan();
}

main()
  .catch((err) => {
    console.error(err);
    process.exitCode = 1;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
